### PR TITLE
Enable user defined display_name for ScalarUDF

### LINKED ISSUE #10376: Support user defined display for UDF
### Is your feature request related to a problem or challenge?

The feature request is based on the need that I would like `get_field(expr, key)` to displayed as `expr[key]`.

two reason
1.  Mimic the pattern as Expr::GetIndexedField for #10374 
2. `expr[key]` looks nicer!

https://github.com/apache/datafusion/blob/8190cb97216e4f46faccbeddae57f6773587955f/datafusion/expr/src/expr.rs#L1583-L1595

Given the current function to build the name has only a fixed pattern, I think enable user-defined display is a good idea

https://github.com/apache/datafusion/blob/8190cb97216e4f46faccbeddae57f6773587955f/datafusion/expr/src/expr.rs#L1646-L1653

### Describe the solution you'd like

_No response_

Adding `display_name` to ScalarUDFImpl

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by yyy1000**
## Which issue does this PR close?

Closes #10376 .

### CODE REVIEW

**Comment by yyy1000** [comment]
> Do you mind adding `display_name` for `get_field` that shows what I expect in the example given in #10376?

Cool, will do that.

**Comment by yyy1000** [comment]
It's good now. @jayzhan211 You can review it when you are available. :)
In this example https://github.com/apache/datafusion/blob/96487ea0cbb7901a1e4aa18fdf6deb8961319fea/datafusion/core/tests/user_defined/user_defined_aggregates.rs#L165-L176, after `apply_function_rewrites`, the logical plan would not create an alias because the `display_name` of `get_field` is the same as `sq.first[value]`

In this PR:
"| initial_logical_plan                                       | Projection: (sq.first)[value], (sq.first)[time]                                                                                                                                                                         |",
    "|                                                            |   SubqueryAlias: sq                                                                                                                                                                                                     |",
    "|                                                            |     Projection: first(t.value,t.time) AS first                                                                                                                                                                          |",
    "|                                                            |       Aggregate: groupBy=[[]], aggr=[[first(t.value, t.time)]]                                                                                                                                                          |",
    "|                                                            |         TableScan: t                                                                                                                                                                                                    |",
    "| logical_plan after apply_function_rewrites                 | Projection: get_field(sq.first, Utf8(\"value\")), get_field(sq.first, Utf8(\"time\"))                                                                                                                                       |",
    "|                                                            |   SubqueryAlias: sq                                                                                                                                                                                                     |",
    "|                                                            |     Projection: first(t.value,t.time) AS first                                                                                                                                                                          |",
    "|                                                            |       Aggregate: groupBy=[[]], aggr=[[first(t.value, t.time)]]                                                                                                                                                          |",
    "|                                                            |         TableScan: t                                                                                                                                                                                                    |",


Before this PR:
"| initial_logical_plan                                       | Projection: (sq.first)[value], (sq.first)[time]                                                                                                                                                                         |",
    "|                                                            |   SubqueryAlias: sq                                                                                                                                                                                                     |",
    "|                                                            |     Projection: first(t.value,t.time) AS first                                                                                                                                                                          |",
    "|                                                            |       Aggregate: groupBy=[[]], aggr=[[first(t.value, t.time)]]                                                                                                                                                          |",
    "|                                                            |         TableScan: t                                                                                                                                                                                                    |",
    "| logical_plan after apply_function_rewrites                 | Projection: get_field(sq.first, Utf8(\"value\")) AS sq.first[value], get_field(sq.first, Utf8(\"time\")) AS sq.first[time]                                                                                                  |",
    "|                                                            |   SubqueryAlias: sq                                                                                                                                                                                                     |",
    "|                                                            |     Projection: first(t.value,t.time) AS first                                                                                                                                                                          |",
    "|                                                            |       Aggregate: groupBy=[[]], aggr=[[first(t.value, t.time)]]                                                                                                                                                          |",
    "|                                                            |         TableScan: t                                                                                                                                                                                                    |",
    "| logical_plan after inline_table_scan                       | SAME TEXT AS ABOVE

File: datafusion/functions/src/core/getfield.rs [hunk 1]
```diff
@@ -50,10 +50,31 @@ impl ScalarUDFImpl for GetFieldFunc {
...
+            _ => {
+                return exec_err!(
+                    "get_field function requires the argument field_name to be a string"
+                );
+            }
+        };
+
+        Ok(format!("{}[{}]", args[0].display_name()?, name))
+    }
+
```
Comment (jayzhan211) [inline datafusion/functions/src/core/getfield.rs:77]: cli
```
DataFusion CLI v38.0.0
> select struct('a');
+-------------------+
| struct(Utf8("a")) |
+-------------------+
| {c0: a}           |
+-------------------+
1 row(s) fetched. 
Elapsed 0.023 seconds.

> select get_field(struct('a'), 'c0');
+-----------------------+
| struct(Utf8("a"))[c0] |
+-----------------------+
| a                     |
+-----------------------+
1 row(s) fetched. 
Elapsed 0.007 seconds.
```

**Review by jayzhan211 — APPROVED** [review]
👍

File: datafusion/functions/src/core/getfield.rs (hunk 1 above)
Comment (yyy1000) [reply datafusion/functions/src/core/getfield.rs:77]: Yeah, this example has better view. 🙌

**Comment by alamb** [comment]
Thanks @yyy1000  and @jayzhan211

**Pull request merged**